### ISSUE TITLE
etcd value to node name translation chops off a character

### ISSUE THREAD

**Issue by alexander-stepanov** [rabbitmq/rabbitmq-peer-discovery-etcd#14]
I'm trying to set up a 3 nodes cluster with RabbitMQ 3.7.7 using the built-in Etcd discovery plugin.

According to the etcdctl (and also seen via the debug RabbitMQ logs), all nodes have successfully registered themselves in the Etcd:

```
# etcdctl ls /nct/co/12.0.4/config/mq/co/nodes
/nct/co/12.0.4/config/mq/co/nodes/rabbit@devops33-2
/nct/co/12.0.4/config/mq/co/nodes/rabbit@devops33-4
/nct/co/12.0.4/config/mq/co/nodes/rabbit@devops33-3
```
Etcd version: 3.2.18
API version: 2

But later in the RabbitMQ logs, I can see that the first character of the peer's hostname is omitted.
For debugging I've modified my /etc/hosts file to include both versions of hostname spelling:

```
[root@devops35-4 ~]# cat /etc/hosts
127.0.0.1   localhost localhost.localdomain localhost4 localhost4.localdomain4
::1         localhost localhost.localdomain localhost6 localhost6.localdomain6
10.100.1.124 devops35-2 devops35-2.dm.com
10.100.1.115 devops35-3 devops35-3.dm.com
10.100.1.127 devops35-4 devops35-4.dm.com
10.100.2.124 evops35-2
10.100.2.115 evops35-3
```

Then I've removed RabbitMQ data directory and re-run the service.
Debug log (skipping the initial part of lager debug messages):
```
2018-08-20 14:53:30.023 [info] <0.33.0> Application lager started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.023 [info] <0.5.0> Log file opened with Lager
2018-08-20 14:53:30.186 [debug] <0.97.0> Lager installed handler lager_backend_throttle into lager_event
2018-08-20 14:53:30.395 [info] <0.33.0> Application jsx started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.395 [info] <0.33.0> Application recon started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.401 [info] <0.33.0> Application os_mon started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.423 [debug] <0.141.0> Supervisor mnesia_sup started mnesia_sup:start_event() at pid <0.142.0>
2018-08-20 14:53:30.428 [debug] <0.141.0> Supervisor mnesia_sup started mnesia_ext_sup:start() at pid <0.143.0>
2018-08-20 14:53:30.429 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_monitor:start() at pid <0.145.0>
2018-08-20 14:53:30.435 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_subscr:start() at pid <0.146.0>
2018-08-20 14:53:30.449 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_locker:start() at pid <0.147.0>
2018-08-20 14:53:30.456 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_recover:start() at pid <0.148.0>
2018-08-20 14:53:30.486 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_tm:start() at pid <0.149.0>
2018-08-20 14:53:30.492 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_checkpoint_sup:start() at pid <0.150.0>
2018-08-20 14:53:30.499 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_controller:start() at pid <0.151.0>
2018-08-20 14:53:30.511 [debug] <0.144.0> Supervisor mnesia_kernel_sup started mnesia_late_loader:start() at pid <0.152.0>
2018-08-20 14:53:30.511 [debug] <0.141.0> Supervisor mnesia_sup started mnesia_kernel_sup:start() at pid <0.144.0>
2018-08-20 14:53:30.511 [info] <0.33.0> Application mnesia started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.511 [info] <0.33.0> Application xmerl started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.527 [debug] <0.158.0> Supervisor inets_sup started ftp_sup:start_link() at pid <0.159.0>
2018-08-20 14:53:30.561 [debug] <0.161.0> Supervisor httpc_profile_sup started httpc_manager:start_link(default, only_session_cookies, ine
ts) at pid <0.162.0>
2018-08-20 14:53:30.561 [debug] <0.160.0> Supervisor httpc_sup started httpc_profile_sup:start_link([{httpc,{default,only_session_cookies}
}]) at pid <0.161.0>
2018-08-20 14:53:30.566 [debug] <0.160.0> Supervisor httpc_sup started httpc_handler_sup:start_link() at pid <0.163.0>
2018-08-20 14:53:30.566 [debug] <0.158.0> Supervisor inets_sup started httpc_sup:start_link([{httpc,{default,only_session_cookies}}]) at p
id <0.160.0>
2018-08-20 14:53:30.572 [debug] <0.158.0> Supervisor inets_sup started httpd_sup:start_link([]) at pid <0.164.0>
2018-08-20 14:53:30.577 [debug] <0.158.0> Supervisor inets_sup started tftp_sup:start_link([]) at pid <0.165.0>
2018-08-20 14:53:30.577 [info] <0.33.0> Application inets started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.578 [info] <0.33.0> Application asn1 started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.578 [info] <0.33.0> Application crypto started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.578 [info] <0.33.0> Application public_key started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.602 [debug] <0.173.0> Supervisor ssl_admin_sup started ssl_pem_cache:start_link([]) at pid <0.174.0>
2018-08-20 14:53:30.613 [debug] <0.173.0> Supervisor ssl_admin_sup started ssl_manager:start_link([]) at pid <0.175.0>
2018-08-20 14:53:30.613 [debug] <0.172.0> Supervisor ssl_sup started ssl_admin_sup:start_link() at pid <0.173.0>
2018-08-20 14:53:30.624 [debug] <0.176.0> Supervisor ssl_connection_sup started tls_connection_sup:start_link() at pid <0.177.0>
2018-08-20 14:53:30.630 [debug] <0.176.0> Supervisor ssl_connection_sup started ssl_listen_tracker_sup:start_link() at pid <0.178.0>
2018-08-20 14:53:30.635 [debug] <0.176.0> Supervisor ssl_connection_sup started dtls_connection_sup:start_link() at pid <0.179.0>
2018-08-20 14:53:30.640 [debug] <0.176.0> Supervisor ssl_connection_sup started dtls_udp_sup:start_link() at pid <0.180.0>
2018-08-20 14:53:30.640 [debug] <0.172.0> Supervisor ssl_sup started ssl_connection_sup:start_link() at pid <0.176.0>
2018-08-20 14:53:30.640 [info] <0.33.0> Application ssl started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.645 [debug] <0.184.0> Supervisor ranch_sup started ranch_server:start_link() at pid <0.185.0>
2018-08-20 14:53:30.645 [info] <0.33.0> Application ranch started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.645 [info] <0.33.0> Application ranch_proxy_protocol started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.645 [info] <0.33.0> Application rabbit_common started on node 'rabbit@devops35-4'
2018-08-20 14:53:30.655 [info] <0.190.0> 
 Starting RabbitMQ 3.7.7 on Erlang 20.3.4
 Copyright (C) 2007-2018 Pivotal Software, Inc.
 Licensed under the MPL.  See http://www.rabbitmq.com/
2018-08-20 14:53:30.676 [info] <0.190.0> 
 node           : rabbit@devops35-4
 home dir       : /root
 config file(s) : /opt/co/config/mq-adv.config
                : /opt/co/config/mq.conf
 cookie hash    : OFFZ242vWRYPjvuYwmA+tQ==
 log(s)         : /logs/rabbit@devops35-4_upgrade.log
                : /logs/rabbitmq.log
 database dir   : /data/rabbit@devops35-4
2018-08-20 14:53:33.131 [debug] <0.191.0> Supervisor rabbit_sup started rabbit_metrics:start_link() at pid <0.194.0>
2018-08-20 14:53:33.132 [debug] <0.195.0> Supervisor rabbit_alarm_sup started rabbit_alarm:start_link() at pid <0.196.0>
2018-08-20 14:53:33.132 [debug] <0.191.0> Supervisor rabbit_sup started rabbit_restartable_sup:start_link(rabbit_alarm_sup, {rabbit_alarm,start_link,[]}, false) at pid <0.195.0>
2018-08-20 14:53:33.135 [info] <0.198.0> Memory high watermark set to 4234 MiB (4440131174 bytes) of 10586 MiB (11100327936 bytes) total
2018-08-20 14:53:33.136 [debug] <0.197.0> Supervisor vm_memory_monitor_sup started vm_memory_monitor:start_link(0.4, #Fun<rabbit_alarm.0.123076882>, #Fun<rabbit_alarm.1.123076882>) at pid <0.198.0>
2018-08-20 14:53:33.136 [debug] <0.191.0> Supervisor rabbit_sup started rabbit_restartable_sup:start_link(vm_memory_monitor_sup, {vm_memory_monitor,start_link,[0.4,#Fun<rabbit_alarm.0.123076882>,#Fun<rabbit_alarm.1.123076882>]}, false) at pid <0.197.0>
2018-08-20 14:53:33.139 [info] <0.200.0> Enabling free disk space monitoring
2018-08-20 14:53:33.139 [info] <0.200.0> Disk free limit set to 50MB
2018-08-20 14:53:33.142 [debug] <0.199.0> Supervisor rabbit_disk_monitor_sup started rabbit_disk_monitor:start_link(50000000) at pid <0.200.0>
2018-08-20 14:53:33.142 [debug] <0.191.0> Supervisor rabbit_sup started rabbit_restartable_sup:start_link(rabbit_disk_monitor_sup, {rabbit_disk_monitor,start_link,[50000000]}, true) at pid <0.199.0>
2018-08-20 14:53:33.142 [info] <0.202.0> Limiting to approx 65436 file handles (58890 sockets)
2018-08-20 14:53:33.142 [debug] <0.201.0> Supervisor file_handle_cache_sup started file_handle_cache:start_link(#Fun<rabbit_alarm.set_alarm.1>, #Fun<rabbit_alarm.clear_alarm.1>) at pid <0.202.0>
2018-08-20 14:53:33.142 [info] <0.203.0> FHC read buffering:  OFF
2018-08-20 14:53:33.143 [debug] <0.191.0> Supervisor rabbit_sup started rabbit_restartable_sup:start_link(file_handle_cache_sup, {file_handle_cache,start_link,[#Fun<rabbit_alarm.set_alarm.1>,#Fun<rabbit_alarm.clear_alarm.1>]}, false) at pid <0.201.0>
2018-08-20 14:53:33.143 [info] <0.203.0> FHC write buffering: ON
2018-08-20 14:53:33.143 [debug] <0.204.0> Supervisor worker_pool_sup started worker_pool:start_link(worker_pool) at pid <0.205.0>
2018-08-20 14:53:33.144 [debug] <0.204.0> Supervisor worker_pool_sup started worker_pool_worker:start_link(worker_pool) at pid <0.206.0>
2018-08-20 14:53:33.144 [debug] <0.204.0> Supervisor worker_pool_sup started worker_pool_worker:start_link(worker_pool) at pid <0.207.0>
2018-08-20 14:53:33.144 [debug] <0.204.0> Supervisor worker_pool_sup started worker_pool_worker:start_link(worker_pool) at pid <0.208.0>
2018-08-20 14:53:33.144 [debug] <0.204.0> Supervisor worker_pool_sup started worker_pool_worker:start_link(worker_pool) at pid <0.209.0>
2018-08-20 14:53:33.144 [debug] <0.191.0> Supervisor rabbit_sup started worker_pool_sup:start_link() at pid <0.204.0>
2018-08-20 14:53:33.144 [info] <0.190.0> Node database directory at /data/rabbit@devops35-4 is empty. Assuming we need to join an existing cluster or initialise from scratch...
2018-08-20 14:53:33.145 [info] <0.190.0> Configured peer discovery backend: rabbit_peer_discovery_etcd
2018-08-20 14:53:33.145 [debug] <0.190.0> Peer discovery backend supports initialisation.
2018-08-20 14:53:33.145 [debug] <0.190.0> Peer discovery etcd: initialising...
2018-08-20 14:53:33.145 [debug] <0.190.0> HTTP client proxy is not configured
2018-08-20 14:53:33.145 [debug] <0.190.0> Peer discovery backend initialisation succeeded.
2018-08-20 14:53:33.145 [info] <0.190.0> Will try to lock with peer discovery backend rabbit_peer_discovery_etcd
2018-08-20 14:53:33.145 [debug] <0.190.0> PUT https://10.100.1.127:2379/v2/keys/nct%2Fco%2F12.0.4%2Fconfig%2Fmq/co/startup_lock?prevExist=false&ttl=30 ["value=rabbit%40devops35-4%20-%20rautjxhydwtnonmlzkebykxauahrllel"]
2018-08-20 14:53:33.186 [debug] <0.190.0> Response: [{ok,{{"HTTP/1.1",201,"Created"},[{"date","Mon, 20 Aug 2018 11:53:33 GMT"},{"content-length","243"},{"content-type","application/json"},{"x-etcd-cluster-id","7de464b1cf58f85f"},{"x-etcd-index","1418818"},{"x-raft-index","8947409"},{"x-raft-term","12"}],"{\"action\":\"create\",\"node\":{\"key\":\"/nct/co/12.0.4/config/mq/co/startup_lock\",\"value\":\"rabbit@devops35-4 - rautjxhydwtnonmlzkebykxauahrllel\",\"expiration\":\"2018-08-20T11:54:03.184677823Z\",\"ttl\":30,\"modifiedIndex\":1418818,\"createdIndex\":1418818}}\n"}}]
2018-08-20 14:53:33.186 [debug] <0.190.0> Starting startup lock refresher
2018-08-20 14:53:33.186 [debug] <0.60.0> Supervisor kernel_safe_sup started timer:start_link() at pid <0.212.0>
2018-08-20 14:53:33.186 [debug] <0.190.0> GET https://10.100.1.127:2379/v2/keys/nct%2Fco%2F12.0.4%2Fconfig%2Fmq/co/nodes?recursive=true
2018-08-20 14:53:33.187 [debug] <0.190.0> Response: {ok,{{"HTTP/1.1",200,"OK"},[{"date","Mon, 20 Aug 2018 11:53:33 GMT"},{"content-length","312"},{"content-type","application/json"},{"x-etcd-cluster-id","7de464b1cf58f85f"},{"x-etcd-index","1418818"},{"x-raft-index","8947409"},{"x-raft-term","12"}],"{\"action\":\"get\",\"node\":{\"key\":\"/nct/co/12.0.4/config/mq/co/nodes\",\"dir\":true,\"nodes\":[{\"key\":\"/nct/co/12.0.4/config/mq/co/nodes/rabbit@devops35-2\",\"value\":\"enabled\",\"expiration\":\"2018-08-20T11:53:58.944379602Z\",\"ttl\":26,\"modifiedIndex\":1418809,\"createdIndex\":1418809}],\"modifiedIndex\":1031,\"createdIndex\":1031}}\n"}}
2018-08-20 14:53:33.187 [info] <0.190.0> All discovered existing cluster peers: rabbit@evops35-2
2018-08-20 14:53:33.187 [info] <0.190.0> Peer nodes we can cluster with: rabbit@evops35-2
2018-08-20 14:53:33.210 [warning] <0.190.0> Could not auto-cluster with node rabbit@evops35-2: {badrpc,nodedown}
2018-08-20 14:53:33.210 [warning] <0.190.0> Could not successfully contact any node of: rabbit@evops35-2 (as in Erlang distribution). Starting as a blank standalone node...
2018-08-20 14:53:33.212 [info] <0.33.0> Application mnesia exited with reason: stopped
```

Meanwhile the tcpdump was capturing network activity:
```
[root@devops35-4 ~]# tcpdump -i any 'port 25672 and dst 10.100.1.124 or dst 10.100.2.124'
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on any, link-type LINUX_SLL (Linux cooked), capture size 262144 bytes
14:53:33.188388 IP devops35-4.51202 > evops35-2.epmd: Flags [S], seq 264931364, win 29200, options [mss 1460,sackOK,TS val 337605672 ecr 0,nop,wscale 7], length 0
```

As you can see, instead of contacting "devops35-2" node (as read from Etcd), it contacted "evops35-2" node.

**Comment by michaelklishin** [comment]
```
2018-08-20 14:53:33.187 [debug] <0.190.0> Response: […] {\"key\":\"/nct/co/12.0.4/config/mq/co/nodes/rabbit@devops35-2\", […]
2018-08-20 14:53:33.187 [info] <0.190.0> All discovered existing cluster peers: rabbit@evops35-2
```

is evidence of a chopped off leading character. Thanks.

**Comment by alexander-stepanov** [comment]
This issue is closely related to setting a custom cluster_formation.etcd.key_prefix in mq.conf.
As a workaround, I've commented it out and the problem went away.

**Comment by michaelklishin** [comment]
@alexander-stepanov ok, thanks, that might be why I couldn't reproduce it so far (my test environment used no prefix).

**Comment by michaelklishin** [comment]
@alexander-stepanov can you please share the peer discovery part of your config file (feel free to edit out anything sensitive). I'd like to use a comparable key prefix.

**Comment by michaelklishin** [comment]
I managed to reproduce. [`get_node_from_key/2`](https://github.com/rabbitmq/rabbitmq-peer-discovery-etcd/blob/master/src/rabbit_peer_discovery_etcd.erl#L207) assumes that etcd prefix and cluster name contain no slashes, which is not always the case.

**Comment by hairyhum** [comment]
@alexander-stepanov the fix is in 3.7.8 release. Can you give it a try?
